File: src/main/java/tech/flandia_yingm/auto_fgo/img/Point.java
package tech.flandia_yingm.auto_fgo.img;

import lombok.Value;
import lombok.var;

@Value
public class Point {

    private final int x, y;

    private final double weight;


    public Point(int x, int y) {
        this(x, y, Double.NaN);
    }

    public Point(int x, int y, double weight) {
        this.x = x;
        this.y = y;
        this.weight = weight;
    }

    public boolean isEmpty() {
        return x == -1 && y == -1 && Double.isNaN(weight);
    }

    public Point add(Point p) {
        return new Point(x + p.x, y + p.y);
    }

    public Point sub(Point p) {
        return new Point(x - p.x, y - p.y);
    }

    public Point multiply(int n) {
        return new Point(x * n, y * n);
    }

    public Point divide(int n) {
        return new Point(x / n, y / n);
    }


    public Point addMultipliedOffset(Point offset, int n) {
        return add(offset.multiply(n));
    }


    public static Point map(Point p,
                            int xStartMin, int xStartMax, int xEndMin, int xEndMax,
                            int yStartMin, int yStartMax, int yEndMin, int yEndMax) {
        var x = p.getX();
        var y = p.getY();
        x = (int) Math.round(((double) (x - xStartMin) / (double) (xStartMax - xStartMin)) * (double) (xEndMax - xEndMin));
        y = (int) Math.round(((double) (y - yStartMin) / (double) (yStartMax - yStartMin)) * (double) (yEndMax - yEndMin));
        return new Point(x, y);
    }

    public static Point getEmpty() {
        return new Point(-1, -1, Double.NaN);
    }

}

File: src/main/java/tech/flandia_yingm/auto_fgo/img/Point.kt
package tech.flandia_yingm.auto_fgo.img

import java.lang.Double.isNaN
import kotlin.math.absoluteValue
import kotlin.math.pow
import kotlin.math.roundToInt
import kotlin.math.sqrt

class Point(
        val x: Int,
        val y: Int,
        val weight: Double = Double.NaN,
        val name: String = ""
) {

    val isValid: Boolean
        get() = x != -1 && y != -1 && !isNaN(weight)


    operator fun plus(p: Point): Point {
        return Point(x + p.x, y + p.y)
    }

    operator fun minus(p: Point): Point {
        return Point(x - p.x, y - p.y)
    }

    operator fun times(n: Int): Point {
        return Point(x * n, y * n)
    }

    operator fun div(n: Int): Point {
        return Point(x / n, y / n)
    }


    override fun toString(): String = if (!weight.isNaN()) "$name($x, $y, $weight)" else "$name($x, $y)"


    companion object {
        val EMPTY: Point
            get() = Point(-1, -1, Double.NaN)

        fun map(p: Point,
                xStartMin: Int, xStartMax: Int, xEndMin: Int, xEndMax: Int,
                yStartMin: Int, yStartMax: Int, yEndMin: Int, yEndMax: Int): Point {
            var x: Int = p.x
            var y: Int = p.y
            x = ((x - xStartMin).toDouble() / (xStartMax - xStartMin).toDouble() * (xEndMax - xEndMin).toDouble()).roundToInt()
            y = ((y - yStartMin).toDouble() / (yStartMax - yStartMin).toDouble() * (yEndMax - yEndMin).toDouble()).roundToInt()
            return Point(x, y)
        }
    }

}

infix fun Point.name(name: String): Point = Point(x, y, weight, name)

fun distance(p1: Point, p2: Point): Double {
    val distPoint = Point((p1 - p2).x.absoluteValue, (p1 - p2).y.absoluteValue)
    return sqrt(distPoint.x.toDouble().pow(2) + distPoint.y.toDouble().pow(2))
}